### DISCUSSION TITLE
[Feature Request] Support for Keyboard Navigation

### DISCUSSION THREAD

**Discussion by FelipeRearden** [chrisgurney/obsidian-note-toolbar#7] [Ideas]
Hello @chrisgurney !!!!

First of all, what an amazing idea ! Thank you very much for creating this plugin 💙

I would like to share with you a FR that could be a game changer for my workflow: **I would like to use the keyboard to navigate and execute commands from the Note Toolbar.**

Let me expand my idea...

1️⃣ A new command to Focus/Unfocus (Toogle) the caret cursor to the Note Toolbar

2️⃣ Keyboard arrow keys <kbd>↔</kbd> and  to navigate between the commands

3️⃣ <kbd>ENTER</kbd> and <kbd>SPACE</kbd> to execute the focused command

<hr>

I hope you like my idea 🙏🙏🙏

Let me know if you need more information !

Thanks for reading this !!!

**Comment by chrisgurney** [comment]
@FelipeRearden Thank you for your feedback! 💙 

I'd love to add keyboard navigation for accessibility. It's high on my list, if I can figure out how to do it well. 😄 

I'm curious if you have further thoughts about:

1. Do you think a command is absolutely required to focus the toolbar? Why not allow the user to shift-tab, or up-arrow to it from the editor? (e.g., try hitting shift-tab when you're in the comment editor here on GitHub; focus goes to the items in the formatting toolbar). 
2. Where should the focus go after executing the item in the toolbar?

**Reply by FelipeRearden** [reply]
> Thank you for your feedback! 💙
> I'd love to add keyboard navigation for accessibility. It's high on my list, if I can figure out how to do it well. 😄

Thank you very much for the reply ! 

> 1. Do you think a command is absolutely required to focus the toolbar? Why not allow the user to back-tab to it from the editor? (e.g., try hitting shift-tab when you're in the comment editor here on GitHub; focus goes to the items in the formatting toolbar).

Yeah :) I have a lot of keyboard hotkeys. This way every time that I have to add a new workflow to my vault I have to manage a new hotkey shortcut. A real example: my shift-tab hotkey opens the core plugin bookmarks modal. 
In my case I would use CRTL-1  to focus on the Note Toolbar.

In a nutshell: a command could make the plugin more versatile for obsidian power users

> 2. Where should the focus go after executing the item in the toolbar?

Great question. After executing the item in the toolbar the focus should go back to active note in the line that the active note was before we execute the commmand to focus on the Note Toolbar.

BUT Here is the magic: before executing the command attached to the item in the Toolbar the first execution should be to go back to the active note line and than execute the  item toolbar command  

There is a native command in Obsidian called `Focus on Last Note` which do this: forces the caret cursor to be active in the active note. 

```
app.commands.executeCommandById('editor:focus')
```

Example: 

1️⃣ The caret cursor is in the active note `NOTEA.md` in `line 11`
2️⃣ Execute the command Focus on Note Toolbar 
3️⃣ <kbd>ENTER</kbd> to  execute the item in the toolbar
4️⃣ Execute Focus on Last Note command (to focus back to the active note)
5️⃣ The caret cursor will go back to NOTEA.md` in `line 11`
6️⃣  Execute the command from the item

In my humble opinion, this way the command triggered from the Note Toolbar will be executed the same way as if was executed directly from a hotkey or the command palette.

<hr> 

Please, let me know if something is not clear about my comments 🙏

**Reply by chrisgurney** [reply]
Great feedback, thank you! Will take your thoughts under advisement (and will probably have clarifying questions).

I'm not sure what your level of experience is, but if you have any tips on how to make the toolbar keyboard navigable from a code perspective, I'm all ears!

So far I've tried:
- setting a `tabindex` on the `div`, and the `li` elements within; and
- adding a `keydown` event listener to the toolbar's container, but it does not appear to be triggered.

I know Obsidian has an event listener registered on the window, which _could_ be skipping the container which I've snuck in between  the properties and the editor... but I'm not sure.

**Reply by FelipeRearden** [reply]
> I'm not sure what your level of experience is, but if you have any tips on how to make the toolbar keyboard navigable from a code perspective, I'm all ears!

Unfortunately I am just an user with almost 4 years of experience using Obsidian running a lot of Templater scripts :) I wish I could understand more about the Obsidian API the share to you more code ideas :)

But some ideas to share:
ℹ️ Maybe the Obsidian API for the Properties Core plugin could help because Properties is 100% keyboard friendly 

ℹ️ On Obsidian Discord server under `plugin-dev` you are gonna find the most smart people on earth. The people who interact there are the Obsidian Legends.  I bet they could help.

ℹ️ Since you are in Beta I bet you could ask for help to joethei or Liam or Marcus.  They are the ones who are in charge of Plugin submission. They are kind people and I bet they could help with this questions. They are experts about plugin development. 

I hope we could find the help that we need for this FR. 🙏 🙏 🙏

**Comment by chrisgurney** [comment]
@FelipeRearden With [v1.3.1](https://github.com/chrisgurney/obsidian-note-toolbar/releases/tag/1.3.1) I've added a `Note Toolbar: Focus` command, which sets focus on the first item. You can then navigate with arrows, and click with enter (or space, as you suggested).

As for adding a command to `unfocus`, in my minimal testing I've noticed:

- Hitting `escape` anywhere in the toolbar returns you to your previous cursor position.
- Clicking on toolbar items should return you to the previous position, as well.

I'd be interested in hearing how it behaves for you. Thanks!

**Reply by FelipeRearden** [reply]
@chrisgurney thank you very much for this new release !!! 

I saw the notification and I am testing for the past 10 minutes :) I am very happy :) 

💡 New command `Note Toolbar: Focus` is working perfect 
💡 The Arrows Navigation are perfect 
💡 Unfocus is perfect as well. In my case, using the command `Focus on Last Note` works the same as `escape` 

I am gonna install on my Main vault and start to stress the whole keyboard navigation with my complex templater commands. I let you know my feedback with a real world example :)

Get back to you in a few hours 🙏
 
<hr> 

As I said before I am in this community in almost 4 years I interact with many developers in these 600 FRs/BRs. 

I just wanna say thank you for work and for this first interaction with me. It is so good to start using a new plugin that is well made and with so many clever details for a first version that I have almost nothing to add from the user workflow standpoint. 

I said almost because I have one more FR to share with you. It is a CSS related FR, but let me play with this amazing plugin before :)

And thank you for your effort to implement my FR. This is a game changer for my workflow 🙏 🙏 🙏

**Reply by chrisgurney** [reply]
Thank you so much for the kind words! It means a lot!

I'll leave this topic open until you've completed your testing.

I did notice a small clipping issue with left/right-aligned items: https://share.cleanshot.com/SQnXNt7BYQRr75NSFTZZ

**Reply by FelipeRearden** [reply]
I finished my stress test and I notice 3 things (about using Note Toolbar from keyboard navigation):

1️⃣ My complex URI commands are running perfect so far 🎉

Nothing to report here. I was afraid about running the commands with text selected in the editor but all good so far.

2️⃣ Editor workspace behavior when using arrow keys (up, down)

When the toolbar is focused I hit arrow UP and DOWN as a instinct . Look the video:

🎥

https://github.com/chrisgurney/obsidian-note-toolbar/assets/85180696/05940604-9f8a-4020-b2f8-dc3bf4fb9bea

My suggestion: lock arrow UP and DOWN to navigate the toolbar items the same way that we have for RIGHT and LEFT . This way the screen will keep “frozen” 

3️⃣ The behavior is perfect on Obsidian Mobile on iPadOS with external mouse and keyboard

**Comment by chrisgurney** [comment]
@FelipeRearden Up and down arrow keys now supported in [v1.3.6](https://github.com/chrisgurney/obsidian-note-toolbar/releases/tag/1.3.6). Thanks again for the suggestion and extensive feedback!

**Reply by FelipeRearden** [reply]
Hello @chrisgurney !!!!

Thank you very much for the new release and the new feature!

Keyboard navigation with UP,DOWN,RIGTH and LEFT keys are working perfect on version `1.3.6` an Obsidian Mobile on iPadOS 🆗

**Comment by chrisgurney** [comment]
Marking as resolved.

**Reply by FelipeRearden** [reply]
100% agree. I get back to this post after BETA to test the behavior in clean installation direct from the community plugins.
